lib/rust/ensogl/core/src/debug/monitor.rs

```diff
@@ -711,143 +704,133 @@ impl PanelData {
         /// Sampler implementation.
-        #[derive(Debug, Default)]
-        pub struct $name {
-            stats: Stats,
-        }
-
-        impl $name {
-            /// Constructor.
-            pub fn new(stats: &Stats) -> Self {
-                Self { stats: stats.clone() }
-            }
-        }
+        #[derive(Copy, Clone, Debug, Default)]
+        pub struct $name {}
 
         impl Sampler for $name {
             fn label(&self) -> &str {
                 $label
             }
-            fn value(&self) -> f64 {
-                let raw_value: f64 = self.stats.$stats_method().as_();
+            fn value(&self, stats: &StatsData) -> f64 {
+                let raw_value: f64 = stats.$stats_field.as_();
                 raw_value / $value_divisor
             }
             fn min_size(&self) -> Option<f64> {
-                Some($t1)
+                Some($warn_threshold)
             }
             fn precision(&self) -> usize {
                 $precision
             }
-            fn check(&self) -> ValueCheck {
-                self.check_by_threshold($t1, $t2)
+            fn check(&self, stats: &StatsData) -> ValueCheck {
+                // Before the first frame, all stat values will be 0, which in (only) this case is
+                // correct.
+                if stats.frame_counter == 0 {
+                    ValueCheck::Correct
+                } else {
+                    let value = self.value(&stats);
+                    ValueCheck::from_threshold($warn_threshold, $err_threshold, value)
+                }
+            }
+            fn max_value(&self) -> Option<f64> {
+                $max_value
             }
         }
     };
 }
 
 const MB: f64 = (1024 * 1024) as f64;
 
-stats_sampler!("Frame time (ms)", FrameTime, frame_time, 1000.0 / 55.0, 1000.0 / 25.0, 2, 1.0);
-stats_sampler!("WASM memory usage (Mb)", WasmMemory, wasm_memory_usage, 50.0, 100.0, 2, MB);
-stats_sampler!("GPU memory usage (Mb)", GpuMemoryUsage, gpu_memory_usage, 100.0, 500.0, 2, MB);
-stats_sampler!("Draw call count", DrawCallCount, draw_call_count, 100.0, 500.0, 0, 1.0);
-stats_sampler!("Buffer count", BufferCount, buffer_count, 100.0, 500.0, 0, 1.0);
-stats_sampler!("Data upload count", DataUploadCount, data_upload_count, 100.0, 500.0, 0, 1.0);
-stats_sampler!("Data upload size (Mb)", DataUploadSize, data_upload_size, 1.0, 10.0, 2, MB);
-stats_sampler!("Sprite system count", SpriteSystemCount, sprite_system_count, 100.0, 500.0, 0, 1.0);
-stats_sampler!("Symbol count", SymbolCount, symbol_count, 100.0, 500.0, 0, 1.0);
-stats_sampler!("Sprite count", SpriteCount, sprite_count, 100_000.0, 500_000.0, 0, 1.0);
-stats_sampler!("Shader count", ShaderCount, shader_count, 100.0, 500.0, 0, 1.0);
+stats_sampler!("Frames per second", Fps, fps, 55.0, 25.0, 2, 1.0, Some(60.0));
+stats_sampler!(
+    "Frame time (ms)",
+    FrameTime,
+    frame_time,
+    1000.0 / 55.0,
+    1000.0 / 25.0,
+    2,
+    1.0,
+    None
```

Comment:
could you make in the macro some params default? Like e.g. if the last param is not passed, use `None`. This way you might be able to create all samplers as one-line expressions :)